dynamicio/core.py

```diff
@@ -286,17 +285,25 @@ def _has_valid_dtypes(self, df: pd.DataFrame) -> bool:
             column_name = col_info.name
             expected_dtype = col_info.data_type
             found_dtype = dtypes[column_name].name
+
             if found_dtype != expected_dtype:
                 if self.show_casting_warnings:
-                    logger.info(f"Expected: '{expected_dtype}' dtype for {self.name}['{column_name}]', found '{found_dtype}'")
+                    logger.info(f"Expected: '{expected_dtype}' dtype for {self.name}['{column_name}'], " f"found '{found_dtype}'")
                 try:
-                    if len(set(type(v) for v in df[column_name].values)) > 1:  # pylint: disable=consider-using-set-comprehension
-                        logger.warning(CASTING_WARNING_MSG.format(column_name, expected_dtype, found_dtype))  # pylint: disable=logging-format-interpolation
-                        logger.info(NOTICE_MSG.format(column_name))  # pylint: disable=logging-format-interpolation
-                    df[column_name] = df[column_name].astype(self.schema.columns[column_name].data_type)
+                    if self.strict_dtype_check:
+                        logger.warning(
+                            f"[strict_dtype_check=True] Performing expensive type-checking for " f"{self.name}['{column_name}']. This may impact performance on large datasets."
+                        )
+                        num_types = df[column_name].map(type).nunique()
+                        if num_types > 1:
+                            logger.warning(CASTING_WARNING_MSG.format(column_name, expected_dtype, found_dtype))
+                            logger.info(NOTICE_MSG.format(column_name))
+
+                    df[column_name] = df[column_name].astype(expected_dtype)
```

Comment:
Previously, type inspection was always performed via `set(type(v) for v in df[column])`, which is inefficient for large `DataFrames`. This patch introduces an early-exit option (`strict_dtype_check=False` by default) that skips the row-wise Python type inspection and directly attempts casting via `astype(...)`. This avoids costly `O(n)` operations per column and makes validation significantly faster for large datasets. When `strict_dtype_check=True`, the logic is preserved with an additional `.map(type).nunique()` for clarity and performance.